python/ray/data/datasource/partitioning.py

```diff
@@ -0,0 +1,378 @@
+    """
+
+    def __init__(
+        self,
+        style: PartitionStyle = PartitionStyle.HIVE,
+        base_dir: Optional[str] = None,
+        field_names: Optional[List[str]] = None,
+        filter_fn: Optional[Callable[[Dict[str, str]], bool]] = None,
+    ):
+        """Creates a new path-based dataset partition parser.
+
+        Args:
+            style: The partition style - may be either HIVE or DIRECTORY.
+            base_dir: "/"-delimited base directory to start searching for partitions
+                (exclusive). File paths outside of this directory will be considered
+                unpartitioned. Specify `None` or an empty string to search for
+                partitions in all file path directories.
+            field_names: The partition key names. Required for DIRECTORY partitioning.
+                Optional for HIVE partitioning. When non-empty, the order and length of
+                partition key field names must match the order and length of partition
+                directories discovered. Partition key field names are not required to
+                exist in the dataset schema.
+            filter_fn: Callback used to filter partitions. Takes a dictionary mapping
+                partition keys to values as input. Unpartitioned files are denoted with
+                an empty input dictionary. Returns `True` to read a file for that
+                partition or `False` to skip it. Partition keys and values are always
+                strings read from the filesystem path. For example, this removes all
+                unpartitioned files:
+                ``lambda d: True if d else False``
+                This raises an assertion error for any unpartitioned file found:
+                ``def do_assert(val, msg):
+                    assert val, msg
+                  lambda d: do_assert(d, "Expected all files to be partitioned!")``
+                And this only reads files from January, 2022 partitions:
+                ``lambda d: d["month"] == "January" and d["year"] == "2022"``
+        """
+        super(PathPartitionParser, self).__init__(style, base_dir, field_names)
+        if style == PartitionStyle.DIRECTORY and not field_names:
+            raise ValueError(
+                "Directory partitioning requires a corresponding list of "
+                "partition key field names. Please retry your request with one "
+                "or more field names specified."
+            )
+        parsers = {
+            PartitionStyle.HIVE: self._parse_hive_path,
+            PartitionStyle.DIRECTORY: self._parse_dir_path,
+        }
+        self._parser_fn: Callable[[str], Dict[str, str]] = parsers.get(style)
+        if self._parser_fn is None:
+            raise ValueError(
+                f"Unsupported partition style: {style}. "
+                f"Supported styles: {parsers.keys()}"
+            )
+        self._filter_fn = filter_fn
+
+    def filter_paths(
```

Comment:
Thinking more about abstractions we are building here, how about we separate this out from Parser as a Filter?
I mean whether this will make sense depends on what you think about the roles of those classes as mentioned above :)

Reply:
I've separated the filter out from the parser in the latest revision. One tangential benefit is the ability to now test the path partition parser independent of partition filtering.